src/hooks/with-auth.tsx

```diff
@@ -0,0 +1,67 @@
+import React, { Component } from 'react';
+import { NextApiRequest, NextApiResponse, NextPage, NextPageContext } from 'next';
+import Router from 'next/router';
+
+import { NullableUserProfile, useUser } from './use-user';
+import { ISignInWithAuth0 } from '../instance';
+import { createLoginUrl } from '../utils/url-helpers';
+
+type RedirectToLoginProps = {
+  render: () => React.ReactElement;
+};
+
+class RedirectToLogin extends Component<RedirectToLoginProps> {
+  public constructor(props: RedirectToLoginProps) {
+    super(props);
+  }
+
+  public componentDidMount(): void {
+    window.location.assign(createLoginUrl(Router.pathname));
+  }
+
+  public render(): React.ReactElement {
+    return this.props.render();
+  }
+}
+
+type AuthenticatedProps = React.PropsWithChildren<{ user: NullableUserProfile }>;
+
+export default function withAuth(
+  InnerComponent: React.ElementType,
+  redirect: () => React.ReactElement,
+  instance: ISignInWithAuth0
+): NextPage<AuthenticatedProps> {
+  const Authenticated: NextPage<AuthenticatedProps> = (props) => {
+    const { user } = useUser();
+
+    if (!user) {
+      return <RedirectToLogin render={redirect} />;
+    }
+
+    return <InnerComponent {...props} user={user} />;
+  };
+
+  Authenticated.getInitialProps = async (context: NextPageContext): Promise<AuthenticatedProps> => {
+    if (!context.req) {
+      const response = await fetch('/api/me');
+      const result = response.ok ? await response.json() : null;
+
+      return { user: result, children: undefined };
+    }
+
+    const session = await instance.getSession(context.req as NextApiRequest, context.res as NextApiResponse);
```

Comment:
`context.req` in `getInitialProps` is an `IncomingMessage`, so we need to update the types od `getSession` to accept `IncomingMessage | NextApiRequest` (and the same for `ServerResponse` and `NextApiResponse`)

Also `getSession` will error if `context.res` is undefined, so we need to figure out when that would be and handle those cases

Reply:
In `next` they're [also casting](https://github.com/vercel/next.js/blob/8277d4d5e305dad03fae6738e8eee07dc686b647/packages/next/next-server/server/next-server.ts#L823), so it seems safe. But it's probably better to use a union type.

Reply:
> Also getSession will error if context.res is undefined, so we need to figure out when that would be and handle those cases
I'll look into that 👍🏼

Reply:
Just tried the union type: `Argument of type 'ServerResponse | undefined' is not assignable to parameter of type 'ServerResponse | NextApiResponse<any>'`. Will have to keep the cast.